src/archive_writer.rs

```diff
@@ -639,32 +1075,78 @@ pub fn write_archive_to_stream<W: Write + Seek>(
         // If LastMemberHeaderOffset isn't going to fit in a 32-bit varible we need
         // to switch to 64-bit. Note that the file can be larger than 4GB as long as
         // the last member starts before the 4GB offset.
-        if last_member_header_offset >= SYM64_THRESHOLD {
+        if maybe_headers_size.unwrap() + last_member_header_offset >= SYM64_THRESHOLD {
             if kind == ArchiveKind::Darwin {
                 kind = ArchiveKind::Darwin64;
             } else {
                 kind = ArchiveKind::Gnu64;
             }
+            maybe_headers_size = None;
         }
     }
 
     if thin {
-        write!(w, "!<thin>\n")?;
+        writeln!(w, "!<thin>")?;
     } else if is_aix_big_archive(kind) {
-        write!(w, "<bigaf>\n")?;
+        writeln!(w, "<bigaf>")?;
     } else {
-        write!(w, "!<arch>\n")?;
+        writeln!(w, "!<arch>")?;
     }
 
+    let headers_size;
     if !is_aix_big_archive(kind) {
-        write_symbol_table(w, kind, &data, &sym_names, 0)?;
+        headers_size = maybe_headers_size.map_or_else(
+            || {
+                compute_headers_size(
+                    kind,
+                    data.len(),
+                    string_table_size,
+                    num_syms,
+                    sym_names.len().try_into().unwrap(),
+                    is_coff_archive(kind).then_some(&sym_map),
+                )
+            },
+            Ok,
+        )?;
+        write_symbol_table(
+            w,
+            kind,
+            &data,
+            &sym_names,
+            headers_size,
+            num_syms,
+            0,
+            0,
+            false,
+        )?;
+
+        if is_coff_archive(kind) {
+            write_symbol_map(w, kind, &data, &sym_map, headers_size)?;
+        }
+
+        if string_table_size > 0 {
+            let string_table_member = string_table_member.unwrap();
+            w.write_all(&string_table_member.header)?;
+            w.write_all(string_table_member.data)?;
+            w.write_all(string_table_member.padding)?;
+        }
+
+        if !sym_map.ec_map.is_empty() {
```

Comment:
LLVM has an additional ShouldWriteSymtab condition. I don't know if that is actually necessary, but I did prefer keeping the divergence with LLVM as small as possible to make reviewing easier.